File: paintWalls.py
from typing import *
from functools import cache

class Solution:
    def paintWalls(self, cost: List[int], time: List[int]) -> int:
      n = len(cost)
      @cache
      def solve(i, rest): 
        if rest <= 0: 
          return 0

        if i == n: 
          return float('inf')
        
        paint = cost[i] + solve(i + 1, rest - 1 - time[idx])
        dontPaint = solve(i + 1, rest)

        return min(paint, dontPaint)

      return solve(1, n)


cost = [1, 2, 3, 2]
time = [1, 2, 3, 2]

sol = Solution()
val = sol.paintWalls(cost, time)
print(val)

File: paintWalls.js
var paintWalls = function (cost, time) {
	const n = cost.length;
	const memo = Array.from({ length: n }, () => Array(n + 1).fill(-1));

	function solve(idx, remaining) {
		if (remaining <= 0) return 0;

		if (idx === n) return Infinity;

		if (memo[idx][remaining] !== -1) return memo[idx][remaining];

		const paint = cost[idx] + solve(idx + 1, remaining - 1 - time[idx]);
		const dontPaint = solve(idx + 1, remaining);

		memo[idx][remaining] = Math.min(paint, dontPaint);
		return memo[idx][remaining];
	}
	return solve(0, cost.length);
};
const cost = [1, 2, 3, 2];
const time = [1, 2, 3, 2];

console.log(paintWalls(cost, time));
